cordova-plugin-outline/apple/vpn/PacketTunnelProvider.m

```diff
@@ -117,55 +114,53 @@ - (void)startTunnelWithOptions:(NSDictionary *)options
-                                        self.connectionStore.isUdpSupported = isUdpSupported;
-                                        self.connectionStore.status = ConnectionStatusConnected;
-
-                                      } else {
-                                        [self execAppCallbackForAction:kActionStart
-                                                             errorCode:vpnPermissionNotGranted];
-                                      }
-                                      completionHandler(error);
-                                    }];
-                         } else {
-                           [self execAppCallbackForAction:kActionStart errorCode:clientErrorCode];
-                           completionHandler([NSError errorWithDomain:NEVPNErrorDomain
-                                                                 code:NEVPNErrorConnectionFailed
-                                                             userInfo:nil]);
-                         }
-                       }];
+  long errorCode = noError;
+  __block NSError *err = nil;
+  if (!isOnDemand) {
+    ShadowsocksCheckConnectivity(self.hostNetworkAddress, [self.connection.port intValue],
+                                 self.connection.password, self.connection.method, &errorCode,
+                                 &err);
+  }
+  __block ErrorCode clientErrorCode =
+      (errorCode == noError || errorCode == udpRelayNotEnabled) ? noError : errorCode;
+  if (clientErrorCode == noError) {
+    [self connectTunnel:[self getTunnelNetworkSettings]
+             completion:^(NSError *error) {
+               if (!error) {
+                 BOOL isUdpSupported =
+                     isOnDemand ? self.connectionStore.isUdpSupported : errorCode == noError;
+                 if (![self startTun2Socks:isUdpSupported]) {
+                   clientErrorCode = vpnStartFailure;
+                   err = [NSError errorWithDomain:NEVPNErrorDomain
+                                             code:NEVPNErrorConnectionFailed
+                                         userInfo:nil];
+                 } else {
+                   [self listenForNetworkChanges];
+                   [self.connectionStore save:connection];
+                   self.connectionStore.isUdpSupported = isUdpSupported;
+                   self.connectionStore.status = ConnectionStatusConnected;
+                 }
+               } else {
+                 clientErrorCode = vpnPermissionNotGranted;
+                 err = [NSError errorWithDomain:NEVPNErrorDomain
+                                           code:NEVPNErrorConnectionFailed
+                                       userInfo:nil];
+               }
+               [self execAppCallbackForAction:kActionStart errorCode:clientErrorCode];
+               completionHandler(err);
+             }];
+  } else {
+    [self execAppCallbackForAction:kActionStart errorCode:clientErrorCode];
```

Comment:
This can be `errorCode` instead of `clientErrorCode`.

Reply:
Done.